webhooks/manualscalertrait_webhook.go

```diff
@@ -0,0 +1,169 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhooks
+
+import (
+	"encoding/json"
+	"fmt"
+
+	logr "github.com/go-logr/logr"
+	adminv1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"net/http"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+
+	"github.com/crossplane/crossplane/apis/oam/v1alpha2"
+)
+
+type ManualScalerTraitValidator struct {
+	Log logr.Logger
+	gvk schema.GroupVersionKind
+}
+
+func (v ManualScalerTraitValidator) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	convertToHttpHandler(v.validate, v.Log)(w, r)
+}
+
+// this is the default way, we will generate the path given gvk
+func (v ManualScalerTraitValidator) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	var err error
+	v.gvk, err = apiutil.GVKForObject(&v1alpha2.ManualScalerTrait{}, mgr.GetScheme())
+	if err != nil {
+		return err
+	}
+	vPath := validate_path_prefix + generatePath(v.gvk)
+	return RegisterWebhookWithManager(mgr, vPath, &v)
+}
+
+func (v ManualScalerTraitValidator) validate(ar adminv1.AdmissionReview) *adminv1.AdmissionResponse {
+	log := v.Log.WithValues("name", ar.Request.Name, "Operation", ar.Request.Operation)
+	log.Info("Start to validate a trait")
+	expectedResource := metav1.GroupVersionResource{Group: "core.oam.dev", Version: "v1alpha2", Resource: "manualscalertraits"}
+
+	if ar.Request.Resource != expectedResource {
+		err := fmt.Errorf("wrong resource, expected %+v, got %+v ", expectedResource, ar.Request.Resource)
+		log.Error(err, "expect resource mismatch")
+		return toErrAdmissionResponse(err, http.StatusBadRequest)
+	}
+
+	raw := ar.Request.Object.Raw
+	msTrait := v1alpha2.ManualScalerTrait{}
+	deserializer := codecs.UniversalDeserializer()
+	if _, _, err := deserializer.Decode(raw, nil, &msTrait); err != nil {
+		log.Error(err, "failed to decode")
+		return toErrAdmissionResponse(err, http.StatusOK)
+	}
+
+	// validate trait spec
+	if msTrait.Spec.ReplicaCount > 10 {
```

Comment:
why add this limit?

Reply:
Make sure that the validator works, I can probably come up with some other way

Reply:
We should remove this limit before merge